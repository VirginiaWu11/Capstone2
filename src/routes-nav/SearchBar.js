import React, { useCallback, useMemo } from "react";
import TextField from "@mui/material/TextField";
import Autocomplete from "@mui/material/Autocomplete";
import CircularProgress from "@mui/material/CircularProgress";
import { useModalContext } from "../coins/ModalContext";
import debounce from "lodash.debounce";
import BackendApi from "../api";
import CoinGeckoApi from "../CoinGeckoApi";

export default function SearchBar() {
  const [open, setOpen] = React.useState(false);
  const [options, setOptions] = React.useState([]);
  const loading = open && options.length === 0;
  const { CoinModal, handleOpen, setClickedCoin } = useModalContext();

  React.useEffect(() => {
    if (!open) {
      setOptions([]);
    }
  }, [open]);

  const onChange = useCallback(
    async (event, value) => {
      if (value) {
        let res = await CoinGeckoApi.getCoins(1, 1, [value.coinGeckoId]);
        setClickedCoin(() => res[0]);
        handleOpen();
      }
    },
    [handleOpen, setClickedCoin]
  );

  const debouncedChangeHandler = useMemo(() => {
    const changeHandler = async (event) => {
      if (event) {
        searchedCoins = await BackendApi.search(
          event.target.value.toLowerCase()
        );
        setOptions([...searchedCoins]);
      }
    };
    return debounce(changeHandler, 300);
  }, []);

  return (
    <>
      <Autocomplete
        onChange={onChange}
        size="small"
        id="asynchronous-demo"
        sx={{ width: 400, m: "auto", backgroundColor: "white" }}
        open={open}
        onOpen={() => {
          setOpen(true);
        }}
        onClose={() => {
          setOpen(false);
        }}
        isOptionEqualToValue={(option, value) => option.name === value.name}
        getOptionLabel={(option) => option.name}
        options={options}
        loading={loading}
        renderInput={(params) => (
          <TextField
            {...params}
            label="Search All Coins"
            onChange={debouncedChangeHandler}
            InputProps={{
              ...params.InputProps,
              endAdornment: (
                <React.Fragment>
                  {loading ? (
                    <CircularProgress color="inherit" size={20} />
                  ) : null}
                  {params.InputProps.endAdornment}
                </React.Fragment>
              ),
            }}
          />
        )}
      />
      <CoinModal />
    </>
  );
}

let searchedCoins = [];
