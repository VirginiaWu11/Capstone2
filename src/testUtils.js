import React from "react";
import { UserContext } from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
};

const UserProvider = ({
  children,
  currentUser = demoUser,
  setCurrentUser,
  token,
  setToken,
  signup,
  signin,
  signout,
  updateProfile,
  infoLoaded,
  setInfoLoaded,
}) => (
  <UserContext.Provider
    value={{
      currentUser,
      setCurrentUser,
      token,
      setToken,
      signup,
      signin,
      signout,
      updateProfile,
      infoLoaded,
      setInfoLoaded,
    }}
  >
    {children}
  </UserContext.Provider>
);

export { UserProvider };
