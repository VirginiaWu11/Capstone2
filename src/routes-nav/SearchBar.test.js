import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import SearchBar from "./SearchBar";
import { UserProvider } from "../testUtils";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <SearchBar />
      </UserProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <SearchBar />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot when logged out", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider currentUser={null}>
        <SearchBar />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
