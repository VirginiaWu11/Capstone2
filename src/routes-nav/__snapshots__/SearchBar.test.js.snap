// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-zsee7j-MuiAutocomplete-root"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="asynchronous-demo"
        id="asynchronous-demo-label"
      >
        Search All Coins
      </label>
      <div
        class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1w8yruj-MuiInputBase-root-MuiFilledInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiFilledInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1u9li0h-MuiInputBase-input-MuiFilledInput-input"
          id="asynchronous-demo"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot when logged out 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-zsee7j-MuiAutocomplete-root"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="asynchronous-demo"
        id="asynchronous-demo-label"
      >
        Search All Coins
      </label>
      <div
        class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1w8yruj-MuiInputBase-root-MuiFilledInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiFilledInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1u9li0h-MuiInputBase-input-MuiFilledInput-input"
          id="asynchronous-demo"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
